tests/unit/test_utils.py

```diff
@@ -40,3 +44,38 @@ def test_run_cmd_logs_stdouterr_on_failure(
         run_cmd(["some", "command"], {})
 
     assert caplog.messages == ['The command "some command" failed', *expect_logs]
+
+
+@mock.patch("subprocess.run")
+@mock.patch("shutil.which")
+@pytest.mark.parametrize("extra_args", [True, False])
+def test_run_cmd_execute_by_abspath(
+    mock_shutil_which: mock.Mock,
+    mock_subprocess_run: mock.Mock,
+    extra_args: bool,
+) -> None:
+    mock_shutil_which.return_value = "/usr/local/bin/yarn"
```

Comment:
Can't say I'm a fan of handling the same type of mock in different ways: `mock_which_go` vs bare `mock_shutil_which.return_value`. At the same time though I'm not sure how the usage of `unittest.mock` would fare in a seemingly more generic solution when it's still mixed up with pytest, so I guess let's leave this as you propose. For the record, here's what I had in mind (uses native pytest's mocking capabilities which work differently):
```
# conftest.py
...
def monkeypatch_context():
    with pytest.MonkeyPatch.context() as mp:
        yield mp

@pytest.fixture(scope="module")
def monkeypatch_module_scope():
    yield from monkeypatch_context()


@pytest.fixture(scope="class")
def monkeypatch_class_scope():
    yield from monkeypatch_context()
...

# unit/test_utils.py
@pytest.fixture(scope="module", autouse=True)
def patch_which_yarn(monkeypatch_module_scope):
    def mock_which(self):
        return "/usr/local/bin/yarn"

    monkeypatch_module_scope.setattr("shutil.which", mock_which)

...
def test_run_cmd_execute_by_abspath(...)
    ...
    # no additional handling of mocks needed, this will already call the patched version of shutil.which 
    run_cmd(
```

^The fixtures defined in `conftest.py` could then be used similarly in different scopes in all modules across the code base, although the way how the `shutil.which` would need to be handled (as showcased) isn't particularly elegant, but hey, at least it's native and the test always calls the patched version by its real name, not `mock_sth_run` :). Let's just leave any of ^this as future work...

Reply:
Or we could introduce flexmock (as future work)...